python/pip_install/private/test/requirements_parser_tests.bzl

```diff
@@ -0,0 +1,185 @@
+
+def _parse_basic_test_impl(ctx):
+    env = unittest.begin(ctx)
+
+    # Base cases
+    asserts.equals(env, [], parse(""))
+    asserts.equals(env, [], parse("\n"))
+
+    # Various requirement specifiers (https://pip.pypa.io/en/stable/reference/requirement-specifiers/#requirement-specifiers)
+    asserts.equals(env, [("SomeProject", "SomeProject")], parse("SomeProject\n"))
+    asserts.equals(env, [("SomeProject", "SomeProject == 1.3")], parse("SomeProject == 1.3\n"))
+    asserts.equals(env, [("SomeProject", "SomeProject >= 1.2, < 2.0")], parse("SomeProject >= 1.2, < 2.0\n"))
+    asserts.equals(env, [("SomeProject", "SomeProject[foo, bar]")], parse("SomeProject[foo, bar]\n"))
+    asserts.equals(env, [("SomeProject", "SomeProject ~= 1.4.2")], parse("SomeProject ~= 1.4.2\n"))
+    asserts.equals(env, [("SomeProject", "SomeProject == 5.4 ; python_version < '3.8'")], parse("SomeProject == 5.4 ; python_version < '3.8'\n"))
+    asserts.equals(env, [("SomeProject", "SomeProject ; sys_platform == 'win32'")], parse("SomeProject ; sys_platform == 'win32'\n"))
+    asserts.equals(env, [("requests", "requests [security] >= 2.8.1, == 2.8.* ; python_version < 2.7")], parse("requests [security] >= 2.8.1, == 2.8.* ; python_version < 2.7\n"))
+
+    # Multiple requirements
+    asserts.equals(env, [("FooProject", "FooProject==1.0.0"), ("BarProject", "BarProject==2.0.0")], parse("""\
+FooProject==1.0.0
+BarProject==2.0.0
+"""))
+
+    asserts.equals(env, [("FooProject", "FooProject==1.0.0"), ("BarProject", "BarProject==2.0.0")], parse("""\
+FooProject==1.0.0
+
+BarProject==2.0.0
+"""))
+
+    # Comments
+    asserts.equals(env, [("SomeProject", "SomeProject")], parse("""\
+# This is a comment
+SomeProject
+"""))
+    asserts.equals(env, [("SomeProject", "SomeProject")], parse("""\
+SomeProject
+# This is a comment
+"""))
+    asserts.equals(env, [("SomeProject", "SomeProject == 1.3 ")], parse("""\
+SomeProject == 1.3 # This is a comment
+"""))
+    asserts.equals(env, [("FooProject", "FooProject==1.0.0"), ("BarProject", "BarProject==2.0.0 ")], parse("""\
+FooProject==1.0.0
+# Comment
+BarProject==2.0.0 #Comment
+"""))
+
+    # Multiline
+    asserts.equals(env, [("certifi", "certifi==2021.10.8     --hash=sha256:78884e7c1d4b00ce3cea67b44566851c4343c120abd683433ce934a68ea58872     --hash=sha256:d62a0163eb4c2344ac042ab2bdf75399a71a2d8c7d47eac2e2ee91b9d6339569")], parse("""\
+certifi==2021.10.8 \
+    --hash=sha256:78884e7c1d4b00ce3cea67b44566851c4343c120abd683433ce934a68ea58872 \
+    --hash=sha256:d62a0163eb4c2344ac042ab2bdf75399a71a2d8c7d47eac2e2ee91b9d6339569
+    # via requests
+"""))
+
+    return unittest.end(env)
+
+def _parse_requirements_lockfile_test_impl(ctx):
+    env = unittest.begin(ctx)
+
+    asserts.equals(env, [
+        ("certifi", "certifi==2021.10.8     --hash=sha256:78884e7c1d4b00ce3cea67b44566851c4343c120abd683433ce934a68ea58872     --hash=sha256:d62a0163eb4c2344ac042ab2bdf75399a71a2d8c7d47eac2e2ee91b9d6339569"),
```

Comment:
do you only need the name for the module extension implementation, and not the hashes? I thought you'd need all the data required to produce a requirements.bzl file, like the hashes for downloading the packages.

OH I remember - the pip_parse rule doesn't actually know how to download anything, it will still end up calling `pip install` on individual packages which are required by the build...

Reply:
Yeah this is the same requirement info that gets passed to `whl_library`.